### ISSUE TITLE
Disable the input action button in the virtual keyboard when TextField is empty

### ISSUE THREAD

**Issue by ryanheise** [flutter/flutter#112198]
## Use case

I'd like to disable the search button in the virtual keyboard when the TextField is empty:

<img src="https://user-images.githubusercontent.com/19899190/191801593-6520d524-cd29-4afc-8632-fa5748764ca5.png" width="300">

## Proposal

On iOS, this can be implemented via [enablesReturnKeyAutomatically](https://developer.apple.com/documentation/uikit/uitextinputtraits/1624432-enablesreturnkeyautomatically?language=objc).

**Comment by huycozy** [comment]
Hi @ryanheise, thanks for filing the issue. 
I see there are some similar open issues addressing the case you described #93915, #30331
Closing this issue as a duplicate, so please follow up on these issues for further updates. You can leave there your comment and react an upvote to the original comment will help to raise the issue. Thank you!

**Comment by ryanheise** [comment]
This feature request is unrelated to #30331 because changing the button label dynamically is a different goal from disabling the button.

As for #93915 , that issue has a different and broader objective of being able to dynamically disable/enable the button at any time, which I'm not sure is even permitted by iOS. Rather, the native iOS API only allows you to follow a single preset behaviour of disabling the button when the text field is empty, and then automatically enabling it when you start typing. See the iOS documentation linked above. My feature request is specifically for that feature.

If I'm wrong, please let me know (i.e. which native API makes #93915 possible). But if #93915 is not actually possible, this present issue still is possible. If we close that issue, then this one should remain open.

**Comment by huycozy** [comment]
Hi @ryanheise 
The reason I bring up the above 2 issues is that I am not sure if a feature when implemented will combine automation and disable/enable together.

With what you described above, I think it matches what the OP of issue #93915 is describing, please see https://github.com/flutter/flutter/issues/93915#issuecomment-974593072. It would be better if similar issues were solved in only one issue, so please leave a comment there for your thoughts and react an upvote to the original comment will help to raise the issue. It will make that issue clearer when implemented.

If you disagree, please write in the comments and I will reopen this issue and continue the discussion. Thanks!

**Comment by ryanheise** [comment]
As before, I don't think #93915 is actually possible. Please correct me if I'm wrong.

**Comment by huycozy** [comment]
Hi @ryanheise I reopen the issue for further discussions.

Why do you think that #93915 is impossible when you are also proposing the similar thing?
- From your proposal:
>I'd like to disable the search button in the virtual keyboard when the TextField is empty:
- From #93915:
>The app is Safari,The Go button is disabled if you enter nothing in the search bar.

---

>please let me know (i.e. which native API makes https://github.com/flutter/flutter/issues/93915 possible)

Also, this is not known in advance, that's the reason why I suggested you should add your comment on that issue so people can see it and give more opinions on it before being implemented (e.g. using [enablesReturnKeyAutomatically](https://developer.apple.com/documentation/uikit/uitextinputtraits/1624432-enablesreturnkeyautomatically?language=objc) as you suggested)

**Comment by ryanheise** [comment]
It sounds like #93915 has proposed a more general feature. If you could have an option to enable/disable the TextField, that would also solve my use case:

```dart
TextField(enableInputAction: true, ...);
```

But it would also solve a broader range of use cases. For example, with that ability, you would also be able to perform validation after each keystroke and disable the action button until the input is valid. You could decide on the condition for when to enable/disable the button, including when the text field is empty. Having such a feature would be nice.

But I'm just proposing a feature that is more narrow:

```dart
TextField(autoEnableReturnKey: true, ...);
```

Basically, I want to expose a native feature of the platform directly.

I assumed it would be best to have a distinct issue for each feature proposal, rather than a distinct issue for each use case, since some feature proposals will solve a broad range of use cases. But I'm fine if you want to handle everything on the other issue. I'll add a comment there.

**Comment by hellohuanlin** [comment]
Using `[enablesReturnKeyAutomatically](https://developer.apple.com/documentation/uikit/uitextinputtraits/1624432-enablesreturnkeyautomatically?language=objc)` would probably work, it's might be more flexible to pass a closure like this: 

```
TextField(enableKeyboardReturnKey: (String text) => text.isNotEmpty, ...)
```

This gives developers full control (e.g. phone number format validation).

**Comment by ryanheise** [comment]
What you're describing is what is already covered by the other issue #93915 so if you like that idea, you should vote for that other issue. Note that using a closure doesn't make any difference one way or the other. You do not need a closure to achieve that because you can simply rebuild the widget tree with a different parameter value as needed.

My proposal here is a different proposal based on just exposing the direct capability of the platform. There's not much we can do about that, that's what the platform supports, so that's what I'm proposing we expose in the API.

**Comment by ryanheise** [comment]
That's true, although with the native API you still don't get any choice about when to disable the button except when the text field is empty so the closure still doesn't help.
